### PR TITLE
[APR] prepare abstract tokenYield query

### PR DESCRIPTION
**Pull request by ribeirojose**
- felt inspired and thought it'd be a cool refactor 😄 
- potentially closes BAL-789 and BAL-797

### CODE REVIEW

File: apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts [hunk 1]
```diff
@@ -20,42 +20,124 @@ export const publicClient = createPublicClient({
...
+      address: "0x1a8F81c256aee9C640e14bB0453ce247ea0DFE6F",
+      token: {
+        address: "0xae78736cd615f374d3085123a210448e74fc6393",
+        symbol: "rETH",
+      },
+    }
   ];
+
+  return data.pool?.priceRateProviders;
 }
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts:99]: I ended up getting to [this](https://bleu-llc.slack.com/archives/C05FE3PLPP0/p1694639017259739) issue with the subgraph and decided not to pursue it further. If the subgraph value is not reliable though we can always fallback to fetching this directly onchain

File: apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts [hunk 2]
```diff
@@ -20,42 +20,122 @@ export const publicClient = createPublicClient({
...
+    blockNumber: blockNumber ? BigInt(blockNumber) : undefined,
+  } as const;
+
+  let rate;
+  try {
+    rate = await publicClient.readContract(args);
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.error(e);
+    rate = -1;
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts:137]: we should log it somewhere when this happens and just default to 0 not to mess with the other numbers

File: apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts [hunk 3]
```diff
@@ -1,8 +1,8 @@
 import { NetworkChainId } from "@bleu-balancer-tools/utils";
-import { createPublicClient, http } from "viem";
+import { createPublicClient, http, zeroAddress } from "viem";
 import { mainnet } from "viem/chains";
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts:3]: here we'll need to select the correct chain for each pool

**Comment by luizakp** [comment]
Cool! will take a closer look tomorrow but seems like BAL-789, BAL-797 are almost done 🥳

**Comment by luizakp** [comment]
note: this is only working for mainnet pool rn

File: apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts [hunk 4]
```diff
@@ -20,43 +20,133 @@ export const publicClient = createPublicClient({
...
-    },
-  ];
+export async function getPoolTokensAprForDate(
+  chain: string,
+  poolId: Address,
+  date: number,
+) {
+  const rateProviders = await getPoolTokensRateProviders(chain, poolId);
+
+  return await Promise.all(
```
Comment (luizakp) [inline apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts:34]: use pThrottle?

Comment (ribeirojose) [reply apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts:34]: why? have you seen this getting rate limited? we won't have many rates here most times

Comment (luizakp) [reply apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts:34]: yup, this is calling multiple times, but not sure why yet

Comment (luizakp) [reply apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts:34]: rn there are some pools that we have 3 tokens at we will do this request 3*number of existing rounds for this pool

**Comment by luizakp** [comment]
example of functioning pool: https://tools-monorepo-git-luiza-bal-784-validate-getrat-01ffd7-bleu-fi.vercel.app/apr/pool/ethereum/0x1e19cf2d73a72ef1332c882f20534b6519be0276000200000000000000000112/round/76

File: apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts (hunk 4 above)
Comment (luizakp) [reply apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts:34]: example of a pool I got rate limited: https://tools-monorepo-git-luiza-bal-784-validate-getrat-01ffd7-bleu-fi.vercel.app/apr/pool/ethereum/0x42ed016f826165c2e5976fe5bc3df540c5ad0af700000000000000000000058b/round/76

Comment (luizakp) [reply apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts:34]: local dev if I retry:
image

**Review by Rawallon — APPROVED** [review]
Looking good, the only

**Pull request merged**